src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal class SourceGeneratedDocumentState : DocumentState
+    {
+        public static SourceGeneratedDocumentState Create(
+            GeneratedSourceResult generatedSourceResult,
+            DocumentId documentId,
+            ISourceGenerator sourceGenerator,
+            HostLanguageServices languageServices,
+            SolutionServices solutionServices)
+        {
+            var textAndVersion = TextAndVersion.Create(generatedSourceResult.SourceText, VersionStamp.Create());
+            ValueSource<TextAndVersion> textSource = new ConstantValueSource<TextAndVersion>(textAndVersion);
+
+            var tree = generatedSourceResult.SyntaxTree;
+
+            // Since the tree is coming directly from the generator, this tree is strongly held so GetRoot() doesn't need a CancellationToken.
+            var root = tree.GetRoot(CancellationToken.None);
+            Contract.ThrowIfNull(languageServices.SyntaxTreeFactory, "We should not have a generated syntax tree for a language that doesn't support trees.");
+
+            if (languageServices.SyntaxTreeFactory.CanCreateRecoverableTree(root))
+            {
+                // We will only create recoverable text if we can create a recoverable tree, otherwise we'd have a
+                // tree wrapping the original text and that may cause identity issues.
+                textSource = CreateRecoverableText(textAndVersion, solutionServices);
+                tree = languageServices.SyntaxTreeFactory.CreateRecoverableTree(
+                    documentId.ProjectId,
+                    filePath: null,
+                    tree.Options,
+                    textSource,
+                    generatedSourceResult.SourceText.Encoding,
+                    root);
+            }
+
+            var treeAndVersion = TreeAndVersion.Create(tree, textAndVersion.Version);
+
+            return new SourceGeneratedDocumentState(
+                languageServices,
+                solutionServices,
+                documentServiceProvider: null,
+                new DocumentInfo.DocumentAttributes(
+                    documentId,
+                    name: generatedSourceResult.HintName,
+                    folders: SpecializedCollections.EmptyReadOnlyList<string>(),
+                    tree.Options.Kind,
+                    filePath: tree.FilePath,
+                    isGenerated: true,
+                    designTimeOnly: false),
+                tree.Options,
+                sourceText: null, // don't strongly hold the text
```

Comment:
i'll be honest here. this sort of thing confuses the heck out of me.   we're passing textSource, but not sourceText (omg thees names).  And we have treeAndVersion, which points at tree, which points at textSource.

I literaly cannot reason about what is holding onto what, and why that may or may not be a problem.  you could literally tell me anything and i would just not and go "i guess so".

Reply:
Yeah, this code (in general) is tricky. A Document object generally under the covers has a _textSource (which can lazily compute the text) and a _treeSource (which can lazily fetch the tree). And yes it's often the case that one source holds onto another source. For example, a fresh Document object for a regular old document has a _textSource that will lazily fetch the file from disk, and then the _treeSource calls the _textSource to get the text and then parses it. In other cases for example if somebody calls WithSyntaxRoot() we might start with a tree and then the text is lazily computed that way if somebody asks for the text.

The sourceText is a funky and odd thing where it seems that in a very specific case when opening a document we hold onto the text strongly because the sources could be deferred or something. I started this whole PR actually trying to delete that insanity and it ended up into some corners I didn't want to touch. I'll see if I can give it a shot maybe next week. 😦